src/components/precision-time-picker.tsx

```diff
@@ -0,0 +1,159 @@
+
+const parseTimeString = (
+  time: string,
+): { hours: number; minutes: number; seconds: number } | null => {
+  if (!validTimeStringRegex.test(time)) {
+    return null;
+  }
+
+  const parts = time.split(':');
+
+  if (parts.length >= 2) {
+    let hours = parseInt(parts[0], 10);
+    hours = hours >= 24 ? 0 : hours;
+    const minutes = parseInt(parts[1], 10);
+    const seconds = parts.length >= 3 ? parseFloat(parts[2]) : 0;
+
+    if (!isNaN(hours) && !isNaN(minutes) && !isNaN(seconds)) {
+      return { hours, minutes, seconds };
+    }
+  }
+
+  return null;
+};
+
+const timeMenuItemsBuilder: () => Array<string> = () => {
+  const items = [];
+
+  for (let hours = 0; hours < 24; hours++) {
+    for (let minutes = 0; minutes < 60; minutes += 30) {
+      items.push(`${padLeadingZero(hours)}:${padLeadingZero(minutes)}`);
+    }
+  }
+
+  return items;
+};
+
+const timeMenuItems = timeMenuItemsBuilder();
+
+export const PrecisionTimePicker: React.FC<PrecisionTimePickerProps> = ({ onChange, time }) => {
+  const [value, setValue] = React.useState(time);
+  const [isValid, setIsvalid] = React.useState(true);
+  const [isMenuOpen, setIsMenuOpen] = React.useState(false);
+  const inputRef = React.useRef<HTMLInputElement | null>(null);
+  const menuRef = React.useRef<HTMLDivElement | null>(null);
+
+  const openMenu = () => {
+    setIsMenuOpen(true);
+  };
+
+  const closeMenu = () => {
+    setIsMenuOpen(false);
+  };
+
+  const handleValueChange = (changedValue: string) => {
+    setValue(changedValue);
+    const parsedTime = parseTimeString(changedValue);
+
+    if (parsedTime) {
+      setIsvalid(true);
+      onChange?.(changedValue, parsedTime.hours, parsedTime.minutes, parsedTime.seconds);
+    } else {
+      setIsvalid(false);
+      onChange?.(changedValue, undefined, undefined, undefined);
+    }
+
+    closeMenu();
+  };
+
+  const onDocumentClick = (event: MouseEvent | undefined) => {
+    if (
+      inputRef.current !== event?.target &&
+      !menuRef.current?.contains(event?.target as HTMLElement)
+    ) {
+      setIsMenuOpen(false);
+    }
+  };
+
+  const handleMenuSelect = (_?: React.MouseEvent, itemId?: string | number) => {
+    if (itemId && typeof itemId === 'string') {
+      handleValueChange(itemId);
+    }
+  };
+
+  const popper = (
+    <Menu
+      id="time-select-menu"
+      onSelect={handleMenuSelect}
+      style={{ maxHeight: '200px', overflow: 'auto' }}
+      ref={menuRef}
+    >
+      <MenuContent>
+        <MenuList>
+          {timeMenuItems.map((item, index) => (
+            <MenuItem
+              key={index}
+              onClick={() => handleValueChange(item)}
+              ref={null}
+              itemID={item}
+              className="precision-time-picker-option"
+            >
+              {item}
+            </MenuItem>
+          ))}
+        </MenuList>
+      </MenuContent>
+    </Menu>
+  );
+
+  const trigger = (
+    <div>
```

Comment:
Do we need to wrap with a `<div>` or can we use `<>...</>`?

Reply:
Yes, using a fragment will make the elements inline in this case and the error text will be on the side of the input rather than below.

Reply:
👍